Java: Improve and add predicates and classes for annotations

Marcono1234 opened the pull request:
Tries to improve the CodeQL classes and predicates for working with annotations.

I have marked this pull request as draft for now to get some feedback whether these changes are desired.
⚠️ Note that I have not tested them yet.

smowton commented on java/ql/src/semmle/code/java/Annotation.qll:
```diff
@@ -44,12 +44,65 @@ class Annotation extends @annotation, Expr {
+   * gets the enum constant used as value for that element. This includes default values in
+   * case no explicit value is specified.
+   */
+  EnumConstant getValueEnumConstant(string name) { result = getValue(name).(FieldRead).getField() }
```
This isn't really specific to annotations. Perhaps provide a class `EnumConstantRead extends FieldRead` with relevant convenience methods?

Marcono1234 replied:
> This isn't really specific to annotations.

It is, isn't it? Enums are one of the value types an annotation type element can have.
Consider this example for [`@Target`](https://docs.oracle.com/en/java/javase/16/docs/api/java.base/java/lang/annotation/Target.html):
```java
@Target(ElementType.ANNOTATION_TYPE)
public @interface MetaAnnotationType {
    ...
}
```
The added CodeQL predicate would allow easily getting `ElementType.ANNOTATION_TYPE` here.

How would a `EnumConstantRead` help for this use case? Or do you mean something different?

smowton replied:
I mean that the desire to interpret a `FieldRead` as an enum constant read goes far beyond annotations, so we should provide something like

```
class EnumConstantRead extends FieldRead {

  EnumConstantRead() { this.getField() instanceof EnumConstant }

  EnumConstant getEnumConstant() { result = this.getField() }

}
```

Then use it in this method.

Marcono1234 replied:
Ah, ok I thought you meant this predicate should be removed.

Though I am not sure if `EnumConstantRead` adds that much value; it probably only saves you one `EnumConstant` cast for the result of `getField()` in most cases.

smowton commented on java/ql/src/semmle/code/java/Annotation.qll:
```diff
@@ -44,12 +44,65 @@ class Annotation extends @annotation, Expr {
+   * gets the string value used for that element. This includes default values in case no
+   * explicit value is specified.
+   */
+  string getValueString(string name) {
```
Rename `getStringValue`, `getIntValue` etc for consistency

Marcono1234 replied:
Your proposed predicate names sound more natural, but I was afraid it would make the CodeQL class more difficult to use. When all predicates for a single value start with `getValue...`, then it is easier to find them and understand their behavior.
When you have `getStringValue`, ... it might be more difficult to tell them apart from `getAValue(name)`.

What do you think?

smowton replied:
Fair point in general, but I think it's better to stick with what we have than introduce a new convention while keeping the old one elsewhere in the library.

Marcono1234 replied:
Have renamed them to  `get...Value`.

smowton commented on java/ql/src/semmle/code/java/Annotation.qll:
```diff
@@ -44,12 +44,65 @@ class Annotation extends @annotation, Expr {
+   * gets the referenced type used as value for that element. This includes default values
+   * in case no explicit value is specified.
+   */
+  Type getValueClass(string name) { result = getValue(name).(TypeLiteral).getReferencedType() }
```
Rename `getReferencedType` for consistency

Marcono1234 replied:
Classes are only one possible value type of annotation type elements, I chose `getValue...` for consistency with the other predicates retrieving an annotation value.

Marcono1234 replied:
Have renamed it to `getTypeValue`.

smowton commented on java/ql/src/semmle/code/java/Annotation.qll:
```diff
@@ -99,18 +166,69 @@ private predicate sourceAnnotValue(Annotation a, Method m, Expr val) {
-  /** Gets an annotation that applies to this element. */
-  Annotation getAnAnnotation() { result.getAnnotatedElement() = this }
+  /**
+   * Gets an annotation that applies to this element, including inherited annotations.
```
This can't be right: it uses `getADeclaredAnnotation` which says it excludes inheritence

Marcono1234 replied:
That should work. Inheritance (as defined by `@Inherited`) only applies to classes, and CodeQL `Class` already overrides this predicate:
https://github.com/github/codeql/blob/f9b539e5b99bfeb72791061146377d5baa6f913e/java/ql/src/semmle/code/java/Type.qll#L632-L640

Marcono1234 replied:
Have added a comment clarifying this.

smowton commented on java/ql/src/semmle/code/java/Annotation.qll:
```diff
@@ -60,16 +113,30 @@ class Annotation extends @annotation, Expr {
+  Expr getAValue(string name) { result = getAValue(name, _) }
+
+  /**
+   * Gets the value at a given index of the annotation element with the specified `name`, which must be
```
The value or a value?

Marcono1234 replied:
Do you mean the mismatch between QLDoc ("the value") and predicate name ("getAValue")?
The predicate name should probably then be `getValue`, but the issue is that this makes the predicate names possibly even more confusing:
- `getAValue()`: Works only for annotations with single value, e.g. `@Target(TYPE)`
- `getValue(string)`: Works only for annotations with single value
- `getAValue(string)`: Works for annotations with single value and with array, e.g. `@Target(TYPE)`, `@Target({TYPE, METHOD)`

Naming the predicate `getValue(string, int)` might make you think at first that it is only for single values (which it is not). But maybe that is not such a big problem because the `int` parameter of the signature already gives a hint that it works for arrays, and you would have the QLDoc explaining its behavior anyways.

Should I change it to `getValue(string, int)` then?

smowton replied:
If the predicate is functional (for a given `name, index` pair there should be either one or zero results) then name it `getValue` and use `the` in the doc. Otherwise use `getAValue` and use `a` in the doc.

Marcono1234 replied:
Do you think it would make sense to deprecate `getAValue(string)` and instead rename it to `getAnArrayValue(string)`?
Then this newly added predicate could be named `getArrayValue(string, int)` which should hopefully make the distinction between the predicates more clear.

Marcono1234 replied:
Have renamed the predicates to `getAnArrayValue(string)` and `getArrayValue(string, int)` and have deprecated the old `getAValue(string)`.

smowton commented on java/ql/src/semmle/code/java/Annotation.qll:
```diff
@@ -99,18 +166,69 @@ private predicate sourceAnnotValue(Annotation a, Method m, Expr val) {
+  private Annotation getAnAssociatedAnnotation(AnnotationType t) {
+    result.getType() = t and
+    // Direct or indirect annotation
+    if getADeclaredAnnotation().getType() = t or getAnIndirectAnnotation().getType() = t
+    then (
+      result = getADeclaredAnnotation() or result = getAnIndirectAnnotation()
```
This looks like it needs an `exists` to insist that the annotation returned by `getADeclaredAnnotation` is the same one reflected in the result

Marcono1234 replied:
Could you please provide a source snippet showing how you would do this?

I think the current implementation is fine (but maybe not very performant?) since the type of the annotation here ultimately does not matter and the result set won't have duplicate elements (due to it being a set). The `if` condition is only needed to make sure that no inherited annotation is part of the result set when a direct or indirect annotation of this type exists.

Though I guess it would be good if I added a comment explaining this.

smowton replied:
Oh, I hadn't noticed the `result.getType() = t` condition is reapplied above, never mind.

smowton replied:
Perhaps this would be easier to read though if you had

```
private Annotation getADeclaredAssociatedAnnotation(AnnotationType t) {
  result.getType() = t and result = [getADeclaredAnnotation(), getAnIndirectAnnotation()]
}

private Annotation getAnAssociatedAnnotation(AnnotationType t) {
  if exists(getADeclaredAssociatedAnnotation(t)) then result = getADeclaredAssociatedAnnotation(t) else ...
}
```

Marcono1234 replied:
Thanks for the suggestion, have applied it!

smowton commented on java/ql/src/semmle/code/java/Annotation.qll:
```diff
@@ -99,18 +166,69 @@ private predicate sourceAnnotValue(Annotation a, Method m, Expr val) {
+      // Only if neither a direct nor an indirect annotation is present look for an inherited one
+      t.isInherited() and
+      // @Inherited only works for classes
+      result = this.(Class).getASupertype().(Class).(Annotatable).getAnAssociatedAnnotation(t)
```
`Class` is always `Annotatable` (via `RefType`)

Marcono1234 replied:
You mean `(Class).(Annotatable)`, right?
The VSCode extension refuses directly using `(Class).getAnAssociatedAnnotation(t)`, I assume because the predicate is private?

smowton replied:
Ah yes, I suppose so, never mind then

smowton commented on java/ql/src/semmle/code/java/Annotation.qll:
```diff
@@ -149,10 +267,54 @@ class AnnotationType extends Interface {
+   */
+  // Note: Cannot use a predicate with string as result because annotation type without
+  // explicit @Target can be applied to all targets, requiring to hardcode element types here
+  bindingset[elementType]
```
If there is a small set of possible values you can do something like `class AnnotationTargetType extends string { AnnotationTargetType() { this = ["CLASS", "METHOD", ...] } }` and drop the `bindingset`

Marcono1234 replied:
Yes that would be doable, but I wanted to avoid this (see "Note: .." comment a few lines above) because the set of possible target types would have to be adjusted then whenever a new constant is added to [`ElementType`](https://docs.oracle.com/en/java/javase/16/docs/api/java.base/java/lang/annotation/ElementType.html) (recents ones were `MODULE` and `RECORD_COMPONENT`).

Additionally it might cause confusing results when a Java database is not using the latest Java version. Consider a Java 8 project for which this predicate claims that `RECORD_COMPONENT` (added for Java 16) is a target.
But maybe that is not actually a big issue, especially since this predicate already relies on JDK-8231435 which was fixed only for Java 14.

I would prefer not using `bindingset` here as well.

smowton commented on java/ql/src/semmle/code/java/JDKAnnotations.qll:
```diff
@@ -19,15 +19,11 @@ class SuppressWarningsAnnotation extends Annotation {
   string getASuppressedWarning() {
-    result = this.getAValue().(StringLiteral).getLiteral() or
-    result = this.getAValue().(ArrayInit).getAnInit().(StringLiteral).getLiteral()
+    result = this.getAValue(_).(CompileTimeConstantExpr).getStringValue()
```
Looks like a use for `getValueString`?

Marcono1234 replied:
`getValueString` only works for single values, it does not work for arrays, which `@SuppressWarnings#value` uses.
I thought about adding `getAValueString` and similar for array values as well, but that would probably make the `Annotation` API too bloated.

Marcono1234 replied:
Have added special array value predicates now (e.g. `getAStringArrayValue(string)`) because otherwise the code accessing the array elements is verbose and error-prone, sometimes erroneously restricting values to `StringLiteral` (or similar) when they could also be a more general `CompileTimeConstantExpr`.

smowton commented on java/ql/src/semmle/code/java/JDKAnnotations.qll:
```diff
@@ -43,10 +39,7 @@ class TargetAnnotation extends Annotation {
-      result = this.getAValue() or
-      result = this.getAValue().(ArrayInit).getAnInit()
-    )
+    result = this.getAValue(_)
```
Keep a no-arg `getAValue` method, since we seem to use it a lot?

Marcono1234 replied:
I did not remove `getAValue()`. But as mentioned in https://github.com/github/codeql/pull/6246/files#r668908394, `getAValue()` and `getAValue(string)` are actually for different use cases. The first one is for single values, the second one is for single and array values.

But while looking at it, I think the `not result instanceof ArrayInit and` in the line above can now be removed with this change.

Marcono1234 replied:
Actually I have marked `getAValue()` as deprecated because it is error-prone. It could by accident match the wrong annotation type element, especially when in the future new elements are added to an annotation type.

smowton reviewed (commented):
Looks broadly good. Three top-level suggestions:

1. Maintain backward compatibility: let the existing `getAnAnnotation` etc retain the current behaviour and introduce new methods that include inheritance
2. Some simple changes (e.g. introducing `getReferencedType`, `getCheckedType`) are noisy and not very related to annotations. Let these have their own PR.
3. Let's have some simple tests to check the right methods do and don't see inheritance

Marcono1234 commented:
> 1. Maintain backward compatibility: let the existing `getAnAnnotation` etc retain the current behaviour and introduce new methods that include inheritance

`getAnAnnotation()` behavior did not change (at least not on purpose). It previously considered inheritance (as defined by `@Inherited`) already, but that is only implemented for CodeQL's `Class` since it only applies to Java classes. The change here is that this documented already for `Annotatable` to make it more clear for the user.

However, `hasAnnotation()` was changed, previously it did not consider inherited annotations. But keeping old behavior (not considering inherited annotations) would probably be confusing because the similarly named predicates `getAnAnnotation()` and `hasAnnotation(string, string)` did and still do consider inherited annotations.

> 2. Some simple changes (e.g. introducing `getReferencedType`, `getCheckedType`) are noisy and not very related to annotations. Let these have their own PR.

Fair point, I will create a separate pull request for them.

smowton commented:
Agreed that `hasAnnotation` deviating from `getAnAnnotation` is likely a bug and we probably don't want to preserve the old behaviour. Deprecating the confusing `getAValue` if it doesn't yield single values also sounds sensible.

Marcono1234 commented:
Sorry for the long delay. Have rebased the changes onto `main` and addressed some of the review comments. The pull request is still not completely ready, but I hope it is going in the right direction.

smowton commented on java/ql/lib/semmle/code/java/Annotation.qll:
```diff
@@ -60,16 +131,78 @@ class Annotation extends @annotation, Expr {
+   * type. This includes default values in case no explicit value is specified.
+   *
+   * If the annotation element is defined with an array initializer, then the result will be one of the
+   * elements of that array. Otherwise, the result will be the single expression defined for the value.
```
```suggestion
   * elements of that array. Otherwise, the result will be the array-typed expression passed for this value.
```
(and similarly all the other predicates duplicating this text)

Marcono1234 replied:
What do you mean by "array-typed"? For example when Java code uses `@Target(ElementType.ANNOTATION_TYPE)`, then `ElementType.ANNOTATION_TYPE` is a single expression for the `ElementType[] value` element. And this predicate will have that single `ElementType.ANNOTATION_TYPE` field read as result.

The "Otherwise, ..." sentence is supposed to cover this exact case where a single expression and no array initializer is used. Maybe I am misunderstanding your comment.

smowton replied:
Ah sure, I thought you wanted to address the case where an array is passed, rather than a value of the array's element type which is implicitly wrapped in an array. Ignore my suggestion.

smowton commented on java/ql/lib/semmle/code/java/Annotation.qll:
```diff
@@ -99,19 +232,73 @@ private predicate sourceAnnotValue(Annotation a, Method m, Expr val) {
+   * The results only include _direct_ annotations; _indirect_ annotations, that is
+   * repeated annotations in an (implicit) container annotation, are not included.
+   */
+  // This predicate is overridden by Class to consider inherited annotations
```
Move comment inside predicate to avoid breaking qldoc apart from the documented predicate

smowton commented on java/ql/lib/semmle/code/java/Annotation.qll:
```diff
@@ -99,19 +232,73 @@ private predicate sourceAnnotValue(Annotation a, Method m, Expr val) {
+  Annotation getADeclaredAnnotation() { result.getAnnotatedElement() = this }
+
+  /** Gets an _indirect_ (= repeated) annotation. */
+  // 'indirect' as defined by https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/reflect/AnnotatedElement.html
```
Move comment inside predicate

smowton commented on java/ql/lib/semmle/code/java/Annotation.qll:
```diff
@@ -99,19 +232,73 @@ private predicate sourceAnnotValue(Annotation a, Method m, Expr val) {
+   * - If an annotation of a type is neither directly nor indirectly present
+   *   the result is an associated inherited annotation (recursively)
+   */
+  // 'associated' as defined by https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/reflect/AnnotatedElement.html
```
Move comment inside predicate

smowton commented on java/ql/lib/semmle/code/java/Annotation.qll:
```diff
@@ -146,10 +333,54 @@ class AnnotationType extends Interface {
-      ann.getAnnotatedElement() = this and
-      ann.getType().hasQualifiedName("java.lang.annotation", "Inherited")
-    )
+    getADeclaredAnnotation().getType().hasQualifiedName("java.lang.annotation", "Inherited")
```
Here and elsewhere: implicit -> explicit `this`

smowton commented on java/ql/lib/semmle/code/java/Annotation.qll:
```diff
@@ -146,10 +333,54 @@ class AnnotationType extends Interface {
+   * enum constants. If no explicit target is specified for this annotation type
+   * it is considered to be applicable to all elements.
+   */
+  // Note: Cannot use a predicate with string as result because annotation type without
```
Move comment inside predicate

smowton commented on java/ql/lib/semmle/code/java/Annotation.qll:
```diff
@@ -146,10 +333,54 @@ class AnnotationType extends Interface {
+    then elementType = getADeclaredAnnotation().(TargetAnnotation).getATargetElementType()
+    else
+      // No Target annotation means "applicable to all contexts" since JDK 14, see https://bugs.openjdk.java.net/browse/JDK-8231435
+      // The compiler does not completely implement that, but pretend it did
```
Expand on this?

Marcono1234 replied:
It think I was referring to some bugs with module declarations, such as [JDK-8254023](https://bugs.openjdk.java.net/browse/JDK-8254023); I will expand the comment.

However, the behavior for a missing `@Target` annotation changed again in Java 17, now it only includes declaration contexts, but excludes type usage contexts ([JDK-8261610](https://bugs.openjdk.java.net/browse/JDK-8261610)). I will mention this in the comment as well.
I will probably still keep the predicate as `isATargetType` for now but exclude `TYPE_USE` if no `@Target` annotation is present, this seems slightly less error-prone than a `getATargetType()` predicate which is more likely to become outdated as the JDK evolves and new declaration contexts are added.

Or what do you think?

smowton replied:
How about naming the predicate `mayBeTargetType`, and note in qldoc that the meaning of no-`@Target` has changed over time, with this predicate taking the broadest interpretation?

Marcono1234 replied:
`mayBeTargetType` (or rather <code>mayBe<ins>A</ins>TargetType</code>?) sounds a bit weird to me, but if you want I can rename it nonetheless.
Though extending the documentation sounds like a good idea, but I think I will rather mention "latest Java version" or similar, since "broadest" might sound like applicable in all contexts (including type contexts), which is no longer the case since Java 17.

Marcono1234 replied:
`getA...` would have to be a predicate with result then, right? That won't be possible with the current implementation.
My main concerns with `mayBeTargetType` is that personally I find that "may be" sounds a bit weird (and too close to "maybe", especially as predicate / method prefix), but this is probably rather subjective; and that this name expresses some uncertainty (I think), even though the predicate is fairly reliable (unless you provide it with strings which are not `ElementType` values).

smowton commented on java/ql/lib/semmle/code/java/Annotation.qll:
```diff
@@ -146,10 +333,54 @@ class AnnotationType extends Interface {
+  }
+
+  /** Holds if this annotation type is annotated with the meta-annotation `@Repeatable`. */
+  predicate isRepeatable() { getADeclaredAnnotation() instanceof RepeatableAnnotation }
```
getADeclared... this isn't inheritable?

Marcono1234 replied:
Yes `getADeclaredAnnotation()` does not include inherited annotations, but `@Repeatable` is not inheritable, so this is correct, isn't it?

smowton replied:
I didn't know, was just asking you to ensure that's what you intended. Sounds like it's fine.

smowton reviewed (commented):
Looks plausible, just some doc fixes here

Marcono1234 commented on java/ql/lib/semmle/code/java/Annotation.qll:
```diff
@@ -44,12 +44,83 @@ class Annotation extends @annotation, Expr {
+  }
+
+  /**
+   * If the annotation element with the specified `name` has a Java `Class` as value type,
```
Have chosen now the following to be consistent with the documentation of the other predicates:
> If the value type of the annotation element with the specified `name` is `java.lang.Class`

I hope that is alright.

Marcono1234 commented:
I think this pull request is now complete and ready for a full review. Thanks already for the previous comments!

Marcono1234 commented on java/ql/lib/change-notes/2022-04-01-annotation-deprecations.md:
```diff
@@ -0,0 +1,7 @@
+* The predicate `Annotation.getAValue()` has been deprecated because it might lead to obtaining the value of the wrong annotation element by accident. `getValue(string)` (or one of the value type specific predicates) should be used to explicitly specify the name of the annotation element.
+* The predicate `Annotation.getAValue(string)` has been renamed to `getAnArrayValue(string)`.
+* The predicate `SuppressWarningsAnnotation.getASuppressedWarningLiteral()` has been deprecated because it unnecessarily restricts the result type; `getASuppressedWarning()` should be used instead.
+* The predicates `TargetAnnotation.getATargetExpression()` and `RetentionAnnotation.getRetentionPolicyExpression()` have been deprecated because getting the enum constant read expression is rarely useful, instead the corresponding predicates for getting the name of the referenced enum constants should be used.
```
Hmm, maybe there are use cases where one wants to get the location of the expression, e.g. to include it in the query result set. But those use cases might still be rare and rather apply to custom annotations and not these JDK default ones.

aschackmull reviewed (approved):
LGTM.

aschackmull commented on java/ql/lib/semmle/code/java/frameworks/spring/SpringController.qll:
```diff
@@ -155,7 +155,7 @@ class SpringRequestMappingMethod extends SpringControllerMethod {
 
   /** Gets the "value" @RequestMapping annotation value, if present. */
   string getValue() {
```
```suggestion
  string getValue() { result = requestMappingAnnotation.getStringValue("value") }
```

aschackmull commented on java/ql/lib/semmle/code/java/frameworks/spring/SpringController.qll:
```diff
@@ -155,7 +155,7 @@ class SpringRequestMappingMethod extends SpringControllerMethod {
   /** Gets the "value" @RequestMapping annotation value, if present. */
   string getValue() {
-    result = requestMappingAnnotation.getValue("value").(CompileTimeConstantExpr).getStringValue()
+    result = requestMappingAnnotation.getStringValue("value")
   }
```
```suggestion
```

aschackmull reviewed (commented):
Missing autoformat.

aschackmull commented:
```
--- expected
+++ actual
@@ -1,1 +1,20 @@
-
+[VALUE_NOT_IN_TYPE] predicate hasLocation(@locatable locatableid, @location id): Value 28381 of field locatableid is not in type @locatable. Appears in tuple (28381,28382)
+	Relevant element: locatableid=28381
+		Full ID for 28381: @"annotation;(27870);(452)". The ID may expand to @"annotation;{@"annotation;{@"field;{@"class;AnnotationValues"};arrayValuesSingleExpr"};{@"class;AnnotationValues$ArrayValues"}"};{@"callable;{@"class;AnnotationValues$ArrayValues"}.annotationValues(){@"array;1;{@"class;AnnotationValues$CustomAnnotation"}"}"}"
+[VALUE_NOT_IN_TYPE] predicate annotValue(@annotation parentid, @method id2, @expr value): Value 28381 of field parentid is not in type @annotation. Appears in tuple (28381,517,28383)
+	Relevant element: parentid=28381
+		Full ID for 28381: @"annotation;(27870);(452)". The ID may expand to @"annotation;{@"annotation;{@"field;{@"class;AnnotationValues"};arrayValuesSingleExpr"};{@"class;AnnotationValues$ArrayValues"}"};{@"callable;{@"class;AnnotationValues$ArrayValues"}.annotationValues(){@"array;1;{@"class;AnnotationValues$CustomAnnotation"}"}"}"
+	Relevant element: id2=517
+		Full ID for 517: @"callable;(448).value()(43)". The ID may expand to @"callable;{@"class;AnnotationValues$CustomAnnotation"}.value(){@"class;java.lang.String"}"
+[VALUE_NOT_IN_TYPE] predicate annotValue(@annotation parentid, @method id2, @expr value): Value 28381 of field value is not in type @expr. Appears in tuple (27870,452,28381)
+	Relevant element: value=28381
+		Full ID for 28381: @"annotation;(27870);(452)". The ID may expand to @"annotation;{@"annotation;{@"field;{@"class;AnnotationValues"};arrayValuesSingleExpr"};{@"class;AnnotationValues$ArrayValues"}"};{@"callable;{@"class;AnnotationValues$ArrayValues"}.annotationValues(){@"array;1;{@"class;AnnotationValues$CustomAnnotation"}"}"}"
+	Relevant element: parentid=27870
+		Full ID for 27870: @"annotation;(27869);(444)". The ID may expand to @"annotation;{@"field;{@"class;AnnotationValues"};arrayValuesSingleExpr"};{@"class;AnnotationValues$ArrayValues"}"
+	Relevant element: id2=452
+		Full ID for 452: @"callable;(444).annotationValues()(449)". The ID may expand to @"callable;{@"class;AnnotationValues$ArrayValues"}.annotationValues(){@"array;1;{@"class;AnnotationValues$CustomAnnotation"}"}"
+[VALUE_NOT_IN_TYPE] predicate exprs(@expr id, int kind, @type typeid, @exprparent parent, int idx): Value 28381 of field parent is not in type @exprparent. Appears in tuple (28383,22,43,28381,-1)
+	Relevant element: parent=28381
+		Full ID for 28381: @"annotation;(27870);(452)". The ID may expand to @"annotation;{@"annotation;{@"field;{@"class;AnnotationValues"};arrayValuesSingleExpr"};{@"class;AnnotationValues$ArrayValues"}"};{@"callable;{@"class;AnnotationValues$ArrayValues"}.annotationValues(){@"array;1;{@"class;AnnotationValues$CustomAnnotation"}"}"}"
+	Relevant element: typeid=43
+		Full ID for 43: @"class;java.lang.String"
[2/8] [36/133 eval 254ms] FAILED(EXTRACTION) /home/runner/work/semmle-code/semmle-code/ql/java/ql/test/library-tests/annotations/DB-CHECK
```

aschackmull commented:
@smowton Looks like we hit an extractor error here.

smowton commented:
I just merged a fix to the Java extractor that resolves the database inconsistency seen above. I've rebased this PR on latest `main` and added a commit amending the test expectations to take the extractor fix into account.

@Marcono1234 I don't think this should impact your work, but the modified extractor records an `annotValue` for nested annotations that *always* uses an array constructor in the case of an annotation-array-typed field. For example, if we have an annotation `@interface Anno1 { Anno2[] nestedAnnotations; }` and a usage `@Anno1(nestedAnnotations = @Anno2("value"))`, then the `annotValue` table will record `Anno1` having a value `{ @Anno2 ... }`, i.e. using an array constructor even though that was implicit at source level. This differs from primitive arrays, where `@interface Anno1 { String[] strings; } ... @Anno1(strings = "value")` would yield two `annotValue` entries, one pointing at the array `{ "value" }` and one pointing at the raw value `"value"`.

Marcono1234 commented:
Thanks! I guess that is fine and I assume you had your reasons for implementing it like this in the extractor.

> would yield two `annotValue` entries, one pointing at the array `{ "value" }` and one pointing at the raw value `"value"`

One would be the source expression and one the non-source expression, right?

Is the "Java Language Tests" check failure due to a failing test, or have your extractor changes simply not been included in the CodeQL version running the "Java Language Tests" yet?

smowton commented:
It failed because the deprecation for `Annotation.getAValue` appears in test output, which I hadn't realised! I'll push a commit to rephrase that test.

smowton reviewed (approved):
Proxying Schack's previous approval
